src/Query.php

```diff
@@ -89,6 +79,17 @@ public function getParameter($name)
         return $this->parameters[$name];
     }
 
+    /**
+     * Returns the http client.
+     *
+     * @return \Piwik\ReportingApi\HttpClient
+     *   The http client.
+     */
+    public function getHttpClient()
```

Comment:
We should remove this method from the `Query` object and instead put it on the `QueryFactory`. Everything that is set here is persisted in subsequent calls, which is not according to how it is intended to be used.

For example:

```
// Set some default parameters on the factory, these are intended to be present in every query.
$query_factory
    ->set('idSite', 1)
    ->set('token_auth', $token);

// Generate a first query.
$query = $query_factory->getQuery('DevicesDetection.getBrowsers');
// Set some options on the query, these should not persist.
$query->setParameter('date', 'today');
$query->getHttpClient()->setMethod('POST');
$query->execute();

// Create a second query.
$query = $query_factory->getQuery('API.getPiwikVersion');
// This should NOT be executed using POST, but it is!
$query->execute();
```

So, since the HttpClient fully persists, it should only be accessible through the `QueryFactory`.

Reply:
Actually I wouldn't make the HttpClient available at all, I don't see the value in it. We could provide proxy methods on the `QueryFactory`: `::setHttpMethod()` and `::getHttpMethod()`.